src/ratelimit_manager.cpp

```diff
@@ -0,0 +1,344 @@
+                        request_counts.threshold_exceed_count_minute = 0;
+                        return true;
+                    }
+                } 
+                else {
+                    return true;
+                }
+            }
+
+
+            auto current_rate_for_hour = (3600 - (std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()) - request_counts.last_reset_time_hour)) / 3600  * request_counts.previous_requests_count_hour;
+            current_rate_for_hour += request_counts.current_requests_count_hour;
+
+            if(rule.max_requests.hour_threshold >= 0 && current_rate_for_hour >= rule.max_requests.hour_threshold) {
+                return true;
+            }
+
+            // Increment request counts
+            request_counts.current_requests_count_minute++;
+            request_counts.current_requests_count_hour++;
+        }
+    }
+
+    return false;
+}
+
+
+
+bool RateLimitManager::allow_entities(RateLimitedEntityType entity_type, const std::vector<std::string> &entities) {
+    std::unique_lock<std::shared_mutex>lock(rate_limit_mutex);
+
+    // Check if a rule exists for the given API keys
+    for(const auto& entity : entities) {
+        if (rate_limit_entities.count(rate_limit_resource_t{entity_type,entity}) > 0) {
+            return false;
+        }
+    }
+
+    // add rate limit for entity
+    rule_store.insert({last_rule_id,rate_limit_rule_t{last_rule_id,RateLimitAction::ALLOW, entity_type,entities, rate_limit_max_requests_t{0, 0}}});
+
+    // Add rule from rule store to rate limit rule pointer
+    for(const auto &entity : entities) {
+        rate_limit_entities.insert({rate_limit_resource_t{entity_type,entity}, &rule_store.at(last_rule_id)});
+    }
+
+    last_rule_id++;;
+    return true;
+}
+
+Option<rate_limit_rule_t> RateLimitManager::find_rule_by_id(const uint64_t id) {
+    std::shared_lock<std::shared_mutex> lock(rate_limit_mutex);
+
+    if(rule_store.count(id) > 0) {
+        return Option<rate_limit_rule_t>(rule_store.at(id));
+    }
+
+
+    return Option<rate_limit_rule_t>(404, "Not Found");
+}
+
+bool RateLimitManager::delete_rule_by_id(const uint64_t id) {
+    std::unique_lock<std::shared_mutex> lock(rate_limit_mutex);
+
+    // Check if a rule exists for the given ID
+    if(rule_store.count(id) > 0) {
+        auto rule = rule_store.at(id);
+
+        // Remove rule from rule store
+        rule_store.erase(id);
+
+        // Remove rule from rate limit rule pointer
+        for(const auto &entity : rule.entity_ids) {
+            rate_limit_entities.erase(rate_limit_resource_t{rule.entity_type,entity});
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+bool RateLimitManager::edit_rule_by_id(const uint64_t id, const rate_limit_rule_t & new_rule) {
+    std::unique_lock<std::shared_mutex> lock(rate_limit_mutex);
+
+    // Check if a rule exists for the given ID
+    if(rule_store.count(id) > 0) {
```

Comment:
Use `rule_store.find()` when you want to do something with the found object. Better than doing a `count` and then an `at` later on.